grpc/shared_conn.go

```diff
@@ -19,70 +25,280 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
+	desc *googlegrpc.StreamDesc,
+	method string,
+	opts ...googlegrpc.CallOption,
+) (googlegrpc.ClientStream, error) {
+	return sc.grpcConn.NewStream(ctx, desc, method, opts...)
 }
 
 // AddOnTrackSub adds an OnTrack subscription for the resource.
 func (sc *SharedConn) AddOnTrackSub(name resource.Name, onTrackCB OnTrackCB) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	sc.resourceOnTrackCBs[name] = onTrackCB
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	sc.resOnTrackCBs[name] = onTrackCB
 }
 
 // RemoveOnTrackSub removes an OnTrack subscription for the resource.
 func (sc *SharedConn) RemoveOnTrackSub(name resource.Name) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	delete(sc.resourceOnTrackCBs, name)
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	delete(sc.resOnTrackCBs, name)
 }
 
-// PeerConn returns the PeerConnection.
+// GrpcConn returns a gRPC capable client connection.
+func (sc *SharedConn) GrpcConn() *ReconfigurableClientConn {
+	return &sc.grpcConn
+}
+
+// PeerConn returns a WebRTC PeerConnection capable of sending video/audio data.  A call to PeerConn
+// concurrent with `ResetConn` or `Close` is allowed to return either the old or new connection.
 func (sc *SharedConn) PeerConn() *webrtc.PeerConnection {
-	return sc.peerConn
+	// Grab a snapshot of the SharedConn state. Release locks before blocking on channel reads.
+	sc.peerConnMu.Lock()
+	readyCh := sc.peerConnReady
+	failCh := sc.peerConnFailed
+	ret := sc.peerConn
+	sc.peerConnMu.Unlock()
+
+	// Block until the peer connection result is known.
+	select {
+	case <-readyCh:
+	case <-failCh:
+	}
+
+	return ret
+}
+
+// ResetConn acts as a constructor for `SharedConn`. ResetConn replaces the underlying
+// connection objects in addition to some other initialization.
+//
+// The first call to `ResetConn` is guaranteed to happen before any access to connection objects
+// happens. But subequent calls can be entirely asynchronous to components/services accessing
+// `SharedConn` for connection objects.
+func (sc *SharedConn) ResetConn(conn rpc.ClientConn, moduleLogger logging.Logger) {
+	sc.grpcConn.ReplaceConn(conn)
+	if sc.logger == nil {
+		// The first call to `ResetConn` happens before anything can access `sc.logger`. So long as
+		// we never write to the member variable, everything can continue to access this without
+		// locks.
+		sc.logger = moduleLogger.Sublogger("conn")
+	}
+
+	if sc.resOnTrackCBs == nil {
+		// Same initilization argument as above with the logger.
+		sc.resOnTrackCBs = make(map[resource.Name]OnTrackCB)
+	}
+
+	sc.peerConnMu.Lock()
+	defer sc.peerConnMu.Unlock()
+
+	if sc.peerConn != nil {
+		sc.logger.Warn("SharedConn is being reset with an active peer connection.")
+		utils.UncheckedError(sc.peerConn.Close())
+		sc.peerConn = nil
+	}
+
+	peerConn, err := webrtc.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
```

Comment:
If we return here won't `readyCh` and `failCh ` block forever?

Reply:
Good catch -- forgot about nil channels.